tiledb/sm/rest/rest_profile.cc

```diff
@@ -128,35 +128,27 @@ void RestProfile::save_to_file() {
   if ((param_values_["rest.username"] == RestProfile::DEFAULT_USERNAME) !=
       (param_values_["rest.password"] == RestProfile::DEFAULT_PASSWORD)) {
     throw RestProfileException(
-        "Failed to save RestProfile; invalid username/password pairing.");
+        "Failed to save; invalid username/password pairing.");
   }
 
   // If the file already exists, load it into a json object.
   json data;
-  std::string original_filepath = filepath_;
   if (std::filesystem::exists(filepath_)) {
-    // Temporarily append filename with a random label to guarantee atomicity.
-    filepath_ = filepath_ + random_label();
-    if (std::rename(original_filepath.c_str(), filepath_.c_str()) != 0) {
-      throw RestProfileException(
-          "Failed to save RestProfile due to an internal error.");
-    }
-
     // Read the file into the json object.
-    {
-      std::ifstream file(filepath_);
-      file >> data;
-    }
+    data = read_file(filepath_);
 
     // If the file is outdated, throw an error. This behavior will evolve.
     if (data["version"] < version_) {
       throw RestProfileException(
           "The version of your local profile.json file is out of date.");
     }
 
-    // If a profile of the given name already exists, remove it.
+    // RestProfiles are immutable, so disallow overwrites.
```

Comment:
Why throw and not replace the profile? Immutable doesn't always mean "cannot change", it also means "you can change parts of it by replacing the whole".

Reply:
re: https://github.com/TileDB-Inc/TileDB/pull/5463#discussion_r1971427475